### DISCUSSION TITLE
YoutubeTV capture with SageTV

### DISCUSSION THREAD

**Discussion by mackandall** [OpenSageTV/Community#66] [Capture Devices]
The following describes one way of getting YouTubeTV from an android tv device into SageTV using a hdmi capture device such as a bmi3000 encoder or a haugepauge hdpvr 2 or colossus 2. I will focus mainly on the bmi1000 network encoder type device here.

Some of the scripts utilized here were refined over at the channelsdvr forum from the crude versions I posited there drawing from what I had picked up on the old sagetv forums. Yes, it is necessary to stress that the original idea and basic concept that made it a possibility originated with SageTV  and the efforts of people like @enternoescape  @fuzzy and many others. 

I will focus solely on how to do this all with  a linux  server and leave it up to others to share their equivalent windows batch scripts as they develop them if so inclined.

Sorry for the formatting with some of the code but I couldn't figure out yet how to get some of the longer code as neat as we used tone able to on the old forums. I will be editing these instructions for clarity as time goes on.

**Hardware:**

We start first with the android capture device. One of the better and, I assume, safer as well as cheaper ones out there is the $20 onn google tv that you can pick up in store at Walmart.

https://www.walmart.com/ip/onn-Google-TV-4K-Streaming-Box-New-2023-4K-UHD-resolution/2835618394?athbdg=L1102&adsRedirect=true

I mainly use some Nvidia shields I have lying around for the purpose, but I have some of the onn devices and can confirm that they are just as good event though not as snappy and the shield tv. Since you would only be using them for capture the snappiness of the interface should not matter, I think. 

For capture purposes I will focus mainly on the hdmi iptv devices that you can typically catch for cheap on eBay depending on the model you want for your particular needs. The devices and their setup are described here:

http://forums.sagetv.com/forums/showthread_t_63177.html?t=63177

You can find them on amazon:

https://www.amazon.com/iseevy-Encoder-Broadcast-Protocols-Facebook/dp/B0831399BL/ref=sr_1_8?crid=VDGJBTAUO4BN&keywords=hdmi+iptv+encoder&qid=1692319147&sprefix=hdmi+iptv+encode%2Caps%2C125&sr=8-8

There are some new 4k versions that do 1080p@60fps easily. The older ones tend to struggle with that and work best at 720p@60fps. Like I said, if you are looking for a bargain for this particular device check eBay. There seems to  always a YouTuber or some other streamer who is either upgrading, or has come into the realization that they are just not the magnetic personality they thought they were.

**STEPS:**

You need to have opendct installed as described here:

https://github.com/enternoescape/opendct

**Configure Android tv device:** 
	
1. You need to install android debugging for linux (or windows if that is what you are using):

https://www.xda-developers.com/install-adb-windows-macos-linux/

2. You should also install Video Lan Client (VLC). In ubuntu linux that would be:
`sudo apt install vlc `

3. Enable usb and network debugging on your android tv device. Here is a guide:

https://www.makeuseof.com/how-to-use-adb-on-android-tv/ 

4. Find out the ip address of you android tv device while it is connected to a monitor or TV and execute the following command:

`adb -s <ip address of device >  connect `

Look at the TV or monitor and grant permissions for usb debugging for your device.

After that do:

`adb devices`

You should see your device listed 
 
Now try this:

`adb -s ip <address of device> shell am start com.google.android.youtube.tvunplugged`

That should start YouTubeTV.

To stop YouTubeTV do:

`adb -s <ip address of device> shell am force-stop com.google.android.youtube.tvunplugged`

**Create Scripts to control android tv device:**

Create the following directory

`sudo mkdir /opt/opendct/SagetvUray`

Then go there with:

`cd /opt/opendct/SagetvUray`

Create the following files with whichever editor you use:

1. contetid.txt  
 
728 fs1hd XdHxqZrpD2U?onboard=1
730 tnthd FVJ2BR5wLlo?onboard=1
731 tbshd cv3L5_6Xsu8?onboard=1
732 usahd 8xN3o2PSL5s?onboard=1

I only have a few examples here. Notice my channel numbers? Your number for each channel should be based on the epg source you are going to use when you configure your guide data. 

2. sageUray0.sh

#!/bin/bash
channel=$1
export var1=“ip address of your android tv device:5555"
sh /opt/opendct/uraytune_central.sh “$channel"

3. sageUray0Stop.sh

#!/bin/bash
IPADD=“ip address for android tv device:5555"
adb -s $IPADD shell input keyevent 86

4. uraytune_central.sh

#!/bin/bash
content_file=“/opt/opendct/SagetvUray/contentid.txt"
content_id=""

if [ -f "$content_file" ]; then
  content_id=$(grep -w "^$1" "$content_file" | cut -d " " -f3)
fi

if [ -z "$content_id" ]; then
  echo "Invalid option or content_id not found in $content_file"
  exit 1
fi
adb -s $var1 shell "am start -a android.intent.action.VIEW -d  https://tv.youtube.com/watch/"$content_id""

This file is based heavily on something a fellow over the Channels dvr forum came up with to streamline the process for hulu capture from android tv devices.

While you are still in the /opt/opendct/SagetvUray/ folder make your .sh file executable with the following two commands:

`sudo chmod +x sageUray0*`

And:

`sudo chmod +x uraytune_central.sh`

**Let us test to make sure  things are working correctly so far.**

Open vlc and go to the streaming url of your network encoder device. To get this we typically go to http://192.168.1.168/ and in the setting you will see where the ts stream is usually something like http://192.168.1.168/0.ts unless you have customized it otherwise.

Go to File>Open Network and then put the ts address in the URL section. You should see feed from your android tv device. 

Go to a terminal and do:

`sh /opt/opendct/SagetvUray/sageUray0.sh 732`

You should see youtubetv fire up and tune the channel. You can kill it with:

`sh /opt/opendct/SagetvUray/sageUray0Stop.sh`

If that worked we are ready for business. 

**Next you are going to configure your network encoder in opendct.** 

I cant seem to get the link to the original instructions from the old page to work so I will go through it. 

Stop both SageTV and opendct. In ubuntu that would be, for SageTV:

`sudo systemctl stop sagetv.service`

And for opendct:

`sudo systemctl stop opendct.service`

Go to the opendct properties file. In ubuntu that would be:

`cd /etc/opendct/conf`

Open the properties file for editing:

`sudo nano opendct.properties`

Or if you want to use a text editor do 

`sudo gedit opendct.properties`

Find the line that reads as follows:

`generic.http.device_names_cs=`

Add a name for your device like so 

`generic.http.device_names_cs=encoder1`

If you have more than one device use commas to separate the devices. I have notices as well that for some reason as opendct stands it doesn’t read the last device in this list so you have to add a placeholder. If I have two devices then I would do:

`generic.http.device_names_cs=uRayA,uRayB,blank`

Save the file and close it.

Now we ware going to run opendct in a terminal so we can see what is going on. Fire up a new terminal and change to the opendct application directory:

`cd /opt/opendct`

Once we are in the folder we do:

`sudo sh console-only`

Let that stay loaded for a minute or two and then press ctl c to kill it. 
Go back to your opendct.properties file and you should see your new devices listed with options. In my case the device was named urayA and so it is  listed after the steps described with the following lines, among others with the same device number:

'sagetv.device.111545164.device_name=urayA
sagetv.device.111545164.stopping_executable=
sagetv.device.111545164.streaming_url=
sagetv.device.111545164.tuning_executable=`

You will now input the streaming url of your encoder device (typically you go to 192.168.1.168 for the device and you will see the url for ts streaming in the settings), the attached tuning executable of for your android tv device, and the stopping executable for you android device. The relevant lines will now look like this:

`sagetv.device.111545164.stopping_executable=/opt/opendct/SagetvUray/sageUray0Stop.sh
sagetv.device.111545164.streaming_url=http\://192.168.1.168/0.ts
sagetv.device.111545164.tuning_executable=/opt/opendct/SagetvUray/sageUray0.sh`

You shouldn’t need to touch any other line really. Save the file once again and run opendct once more from the terminal as just described. We are doing it this way for now in case the opendct.properties file needs editing. If it does we would kill the process in the terminal again with “ctl c” before editing that file.

Restart Sagetv (in different terminal window)

`sudo systemctl start sagetv.service`

Go into the section where you add new  devices. You should see your device as an option. Configure it with Schedules Direct or which ever other provider you use for epg. Just remember that the channel numbers have to be consistent with what is in the contentid.txt file. 

When you are done and have guide data you should be able to tune live tv and switch between channels.  If all is well then you can kill the opendct terminal once more and restart it properly with:

`sudo systemctl start opendct.service`

**Comment by Sagetvmaster** [comment]
Thanks, but that sure is more complicated than a set-top or satellite box, isn’t it.  Hollywood must be so happy they’ve nearly stopped us evil people who pay them to watch their shows at our convenience.  How dare we…lol

**Reply by mackandall** [reply]
I know, right? 

I could reduce the length with links to the old forums for setting up opendct and the network encoder.  That might make the process not seem so daunting. I decided to describe the steps in detail here since some links to them on the old  forum don't seem to work. 

There is another way of doing this by using the hdmi-for-channels application that has a web interface for setting up all this and feeding a m3u into channels dvr. 

https://github.com/sullrich/ah4c

SageTv could be made to leverage that m3u feed in a way similar to what I described in the old forums for channelsdvr and sagetv integration.  My goal here though was to describe how to do it without using any external applications.

**Comment by KryptoNyte-Sage** [comment]
I'm a Windows user, but I appreciate your work here, and the time it took to lay that out.  Hopefully a Windows user will come in and translate.

Thanks!

**Comment by mackandall** [comment]
https://github.com/sullrich/ah4c

The m3u generated by this  works with sagetv for an easier method than the one described above.
